docs/source/checkpoints/dist_ckpt.rst

```diff
@@ -0,0 +1,392 @@
+Additionally, as shown in the example above, some wrappers are provided that allow to parallelize the save and load across the whole workload (assuming some data duplication).
+
+For "common" strategies, currently the only supported one is `torch` which saves "common" data into a `common.pt` file.
+
+Optimizers
+==========
+This module gives helper tools to the user to simplify constructing ShardedTensors for optimizer states.
+The ShardedTensors that define local to sharded tensors mapping for model parameters should be reused for optimizer states to avoid code duplication.
+
+To this end, the `dist_checkpointing.optimizers.get_param_id_to_sharded_param_map` function can build a mapping between optimizer params ids and model ShardedTensors.
+This mapping can be used by the `dist_checkpointing.optimizers.optim_state_to_sharding_state` function or application code (for non-standard use cases) to construct optimizer sharded state dict with ShardedTensors.
+This should support most optimizer cases, but some of them might require custom sharded state dict creation.
+A good example is a Distributed Optimizer which flattens the parameters - see `Tensors transformations`_ section for more details.
+
+Note: in order to reuse model SharderTensors to create optimizer ShardedTensors, the model **SharderTensors must wrap model parameters**, not just tensors
+(obtaining a state dict with model parameters can be achieved by passing `keep_vars=True` to the model `state_dict` function).
+Otherwise the correspondence between model ShardedTensors and optimizer states is impossible to recreate.
+This is the reason for introducing ShardedTensorFactories - we have to register the original model parameter as `ShardedTensorFactories.data` and apply any subsequent transformations as a factory function in order to make sure that the same transformation can be applied to the optimizer states.
+
+
+Tensors transformations
+=======================
+ShardedTensor API allows declaring some basic transformations that should be performed on the tensors during saving and loading.
+
+Shape mismatch
+--------------
+The `allow_shape_mismatch` flag allows to relax the requirement of matching global tensor shapes during loading.
+Extra padding is filled with zeros or stripped depending on the mismatch kind.
+This comes handy for layers like embedding which might be padded according to parallelism for performance reasons.
+
+Flattening
+----------
+The `flattened_range` attribute allows to declare the fact that `ShardedTensor.data` is actually a slice of a flattened model parameter.
+This corresponds to a transformation used in Distributed Optimizers which flattens the data and shards it along the data-parallel domain.
+
+Extra flattening comes with an efficiency challenge during checkpoint resharding.
+Since flattening is applied after the global tensors is sharded into the grid of local chunks, loading after resharding requires accessing incontiguous data fragments.
```

Comment:
Because flattening occurs after sharding the global tensor into a grid of local chunks, loading after reshaping necessitates accessing non-contiguous data fragments.